### PR TITLE
NsxAlbInfraSetting validations and configuring SEGroup during ingestion

### PR DESCRIPTION
**Pull request by chauhanshubham**
On creates/updates to the NsxAlbInfraSetting object, AKO would configure the SEGroup
in Avi with appropriate clustername labels. In case of deletes, we need to remove the
configured SEGroup labels, in case no other infraSetting CRD is using the SEGroup.
Introduces cache for infraSetting to seGroup Names to track seGroup changes.
This also changes SegGroup references in NsxAlbInfraSetting struct to SeGroup.
Adds configuration steps (both addition and removal of labels) for SeGroups during reboots.

This introduces NsxAlbInfraSetting Unit tests as well.

### CODE REVIEW

**Review by sudswasavi — COMMENTED** [review]
Will wait for the UTs PR post this, few comments inline. @parimanur / @monotosh-avi / @akshayhavile PTAL

File: internal/cache/controller_obj_cache.go [hunk 1]
```diff
@@ -2355,54 +2355,80 @@ func checkRequiredValuesYaml() bool {
...
 		utils.AviLog.Warnf("Service Engine Group: serviceEngineGroupName not set in values.yaml, skipping sync.")
 		return false
 	}
-	uri := "/api/serviceenginegroup/?include_name&name=" + segName + "&cloud_ref.name=" + utils.CloudName
+
+	// List all service engine groups, for every SEG check for NsxAlbInfraSettings,
+	// if NsxAlbInfraSetting NOT found, remove label if exists,
+	// if NsxAlbInfraSetting found, configure label if doesn't exist.
+	// This takes care of syncing SeGroup label settings during reboots.
+	uri := "/api/serviceenginegroup/?include_name&cloud_ref.name=" + utils.CloudName
```
[the remark this answers is not in the document]
Comment (parimanur) [reply internal/cache/controller_obj_cache.go:2378]: Should we also add page size?

File: internal/cache/controller_obj_cache.go [hunk 2]
```diff
@@ -2431,8 +2457,70 @@ func checkSegroupLabels(client *clients.AviClient) bool {
...
 
+	labels := seGroup.Labels
+	if len(labels) == 0 {
+		return true
+	}
+
+	uri := "/api/serviceenginegroup/" + *seGroup.UUID
+	seGroup.Labels = []*models.KeyValue{}
+	response := models.ServiceEngineGroupAPIResponse{}
+	// If tenants per cluster is enabled then the X-Avi-Tenant needs to be set to admin for vrfcontext and segroup updates
```
[the remark this answers is not in the document]
Comment (parimanur) [reply internal/cache/controller_obj_cache.go:2476]: @sudswasavi Should work as write permission is there for SeGroup.

File: internal/nodes/avi_model_l7_hostname_shard.go [hunk 3]
```diff
@@ -358,9 +358,8 @@ func sniNodeHostName(routeIgrObj RouteIngressModel, tlssetting TlsSettings, ingN
 				certsBuilt = true
 			}
 		}
-		if lib.GetSEGName() != lib.DEFAULT_SE_GROUP {
-			sniNode.ServiceEngineGroup = lib.GetSEGName()
-		}
+
+		sniNode.ServiceEngineGroup = lib.GetSEGName()
```
Comment (parimanur) [inline internal/nodes/avi_model_l7_hostname_shard.go:362]: Can you also check for same block of code for EVH? Rebase, as the code is merged yesterday.

File: internal/nodes/avi_model_l7_translator.go [hunk 4]
```diff
@@ -304,11 +303,14 @@ func (o *AviObjectGraph) ConstructAviL7VsNode(vsName string, key string) *AviVsN
...
 
 	// This is a shared VS - always created in the admin namespace for now.
-	avi_vs_meta = &AviVsNode{Name: vsName, Tenant: lib.GetTenant(),
-		EastWest: false, SharedVS: true}
-	if lib.GetSEGName() != lib.DEFAULT_SE_GROUP {
-		avi_vs_meta.ServiceEngineGroup = lib.GetSEGName()
+	avi_vs_meta = &AviVsNode{
+		Name:               vsName,
+		Tenant:             lib.GetTenant(),
+		EastWest:           false,
```
Comment (parimanur) [inline internal/nodes/avi_model_l7_translator.go:309]: this can be ignored.

File: internal/cache/controller_obj_cache.go [hunk 5]
```diff
@@ -2431,8 +2457,70 @@ func checkSegroupLabels(client *clients.AviClient) bool {
 	return true
 }
 
-func checkTenant(client *clients.AviClient) bool {
+// RemoveSeGroupLabels removes labels on the SeGroup
+func RemoveSeGroupLabels(client *clients.AviClient, seGroup *models.ServiceEngineGroup) bool {
```
Comment (parimanur) [inline internal/cache/controller_obj_cache.go:2461]: This behaviour needs to be validated with the SE team, If the existing routes are there for the SEG then those will not be removed from SE if we were to delete/change the labels.

File: internal/cache/controller_obj_cache.go [hunk 6]
```diff
@@ -2355,54 +2355,80 @@ func checkRequiredValuesYaml() bool {
...
 
-	if seg.UUID == nil {
-		utils.AviLog.Warnf("Failed to get UUID for Service Engine Group: %s", segName)
-		return false
+	for _, elem := range elems {
+		seg := models.ServiceEngineGroup{}
+		err = json.Unmarshal(elem, &seg)
+		if err != nil {
+			utils.AviLog.Warnf("Failed to unmarshal data, err: %v", err)
+			continue
```
[the remark this answers is not in the document]
Comment (chauhanshubham) [reply internal/cache/controller_obj_cache.go:2407]: I think continue is fine, I intended to configure as many SeGroups as possible.

**Comment by chauhanshubham** [comment]
build ako

**Comment by sudswasavi** [comment]
I think we have to introduce ServiceType as well on the InfraCRD because for serviceType: `NodePortLocal` - the SE Group labelling is not necessary. @parimanur / @monotosh-avi thoughts?

**Comment by monotosh-avi** [comment]
> I think we have to introduce ServiceType as well on the InfraCRD because for serviceType: `NodePortLocal` - the SE Group labelling is not necessary. @parimanur / @monotosh-avi thoughts?

@sudswasavi It would be nice to have serviceType as part of NsxAlbInfraL7Settings but I think it may require significant changes because of all the mappings we maintain for different serviceTypes.

**Comment by sudswasavi** [comment]
> > I think we have to introduce ServiceType as well on the InfraCRD because for serviceType: `NodePortLocal` - the SE Group labelling is not necessary. @parimanur / @monotosh-avi thoughts?
> 
> @sudswasavi It would be nice to have serviceType as part of NsxAlbInfraL7Settings but I think it may require significant changes because of all the mappings we maintain for different serviceTypes.

If you don't do that, how would you know if you should label the SE or not? We could alternatively assume that the entire cluster is either NodePortLocal or ClusterIP or NodePort by just looking up the global flag for it. Which means, if it's NodePort or NodePortLocal, we can assume it's the same for the CRD too and that way stop labelling the SE groups.

**Comment by monotosh-avi** [comment]
> > > I think we have to introduce ServiceType as well on the InfraCRD because for serviceType: `NodePortLocal` - the SE Group labelling is not necessary. @parimanur / @monotosh-avi thoughts?
> > 
> > 
> > @sudswasavi It would be nice to have serviceType as part of NsxAlbInfraL7Settings but I think it may require significant changes because of all the mappings we maintain for different serviceTypes.
> 
> If you don't do that, how would you know if you should label the SE or not? We could alternatively assume that the entire cluster is either NodePortLocal or ClusterIP or NodePort by just looking up the global flag for it. Which means, if it's NodePort or NodePortLocal, we can assume it's the same for the CRD too and that way stop labelling the SE groups.

Yeah, we have to read from env variable in the case. Anyway let's discuss about this - what would it take to support service type in CRD.

**Review by monotosh-avi — APPROVED** [review]

**Review by parimanur — APPROVED** [review]

**Pull request merged**